WooCommerce/Classes/ViewRelated/Coupons/CouponManagementViewController.swift

```diff
@@ -1,25 +1,205 @@
+            emptyStateViewController.view.trailingAnchor.constraint(equalTo: view.trailingAnchor),
+            emptyStateViewController.view.topAnchor.constraint(equalTo: view.topAnchor),
+            emptyStateViewController.view.bottomAnchor.constraint(equalTo: view.bottomAnchor)
+        ])
+        emptyStateViewController.didMove(toParent: self)
+    }
+
+    /// Removes EmptyStateViewController child view controller if applicable.
+    ///
+    func removeNoResultsOverlay() {
+        guard let emptyStateViewController = emptyStateViewController,
+              emptyStateViewController.parent == self
+        else { return }
+
+        emptyStateViewController.willMove(toParent: nil)
+        emptyStateViewController.view.removeFromSuperview()
+        emptyStateViewController.removeFromParent()
+        self.emptyStateViewController = nil
+    }
+}
+
+
+// MARK: - TableView Data Source
+//
+extension CouponManagementViewController: UITableViewDataSource {
+
+    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
+        viewModel.couponViewModels.count
+    }
+
+    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+        let cell = tableView.dequeueReusableCell(withIdentifier: TitleBodyTableViewCell.reuseIdentifier, for: indexPath)
+        if let cellViewModel = viewModel.couponViewModels[safe: indexPath.row] {
+            configure(cell as? TitleBodyTableViewCell, with: cellViewModel)
+        }
+
+        return cell
+    }
+
+    func configure(_ cell: TitleBodyTableViewCell?, with cellViewModel: CouponListCellViewModel) {
+        cell?.titleLabel.text = cellViewModel.title
+        cell?.bodyLabel.text = cellViewModel.subtitle
+        cell?.accessibilityLabel = cellViewModel.accessiblityLabel
+    }
+
+}
+
+
+// MARK: - Nested Types
+//
+private extension CouponManagementViewController {
+
+    enum Constants {
+        static let estimatedRowHeight = CGFloat(86)
+        static let placeholderRowsPerSection = [3]
+    }
+
+}
+
+
 // MARK: - Localization
 //
 private extension CouponManagementViewController {
+
     enum Localization {
-        static let title = NSLocalizedString("Coupons", comment: "Coupon management coupon list screen title")
+        static let title = NSLocalizedString(
+            "Coupons",
+            comment: "Coupon management coupon list screen title")
+
+        static let emptyStateMessage = NSLocalizedString(
+            "No coupons yet",
+            comment: "The text on the placeholder overlay when there are no coupons on the coupon management list")
+
+        static let emptyStateDetails = NSLocalizedString(
+            "Market your products by adding a coupon to offer your customers a discount.",
+            comment: "The details on the placeholder overlay when there are no coupons on the coupon management list")
     }
+
```

Comment:
very nit: here I would prefer 2 braces without newline between. And there are more cases of similar spaced closing braces above.

I've spend some time to find a rule about it and I think it aligns with https://github.com/raywenderlich/swift-style-guide/#spacing:

> There should be no blank lines after an opening brace or before a closing brace.

(don't assume strict style requirements, I've just noticed inconsistency between VC and VM looking side-by-side 🙂)

Reply:
Yeah, the inconsistency wasn't great. I've removed the excess newlines and prefer it without them too. 

I think I've taken different cues from different existing files as I've been working, but hadn't spotted that in the RW style guide. Thanks for finding that 🙇 it's a well stated guideline, nice and clear.